# 1. Interpretation and significance

The model tells us that the bias term is 873 USD. As a families annual income increases, their expenditure also increases at a rate of .0012 * (annual income) and .00002 (annual income squared). However, if the family also has children, their recreation expenditure decreases by about 223 USD flat. 

The annual income sqaured will have more impact as the income increases--it will increase exponentially.

Since we do not know what the estimated pvalues are, we are not certain if the estimated coefficients are statistically significant, or if just noise (and shoule be treated as zero).

# 2. Weather Model

## Imports and DataFrame Creation

In [0]:
#Arrays and Dataframe
import numpy as np
import pandas as pd

#SQL
from sqlalchemy import create_engine

#Visualization
from matplotlib import pyplot as plt
import seaborn as sns
sns.set()

#Data Exploration
from scipy import stats

#Data Modeling
from sklearn import linear_model
from sklearn.svm import LinearSVC
import statsmodels.api as sm
from sklearn.model_selection import train_test_split
from statsmodels.stats.outliers_influence import variance_inflation_factor 
from statsmodels.tools.tools import add_constant

from statsmodels.tsa.stattools import acf

from scipy.stats import jarque_bera
from scipy.stats import normaltest

In [2]:
postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'weatherinszeged'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

df = pd.read_sql_query('select * from weatherinszeged', con=engine)

# no need for an open connection, as we're only doing a single query
engine.dispose()

/usr/local/lib/python3.6/dist-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)


## Linear Regression Model

In [0]:
#create new variable
df['temp_diff'] = df.apparenttemperature - df.temperature
y = df['temp_diff']
X = df[['humidity', 'windspeed']]

In [0]:
#split the data
X_train, X_test, y_train, y_test = train_test_split(X, y)

In [18]:
#Linear Regression using stats model
# We need to manually add a constant
# in statsmodels' sm
X = sm.add_constant(X)

results = sm.OLS(y, X).fit()

results.summary()

/usr/local/lib/python3.6/dist-packages/numpy/core/fromnumeric.py:2495: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              temp_diff   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Tue, 05 Nov 2019   Prob (F-statistic):               0.00
Time:                        13:56:10   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.4381      0.021    115.948      0.000       2.397       2.479
humidity      -3.0292      0.024   -126.479      0.000      -3.076      -2.982
windspeed     -0.1193      0.001   -176.164      0.000      -0.121      -0.118
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.267
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                          -0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The low pvalue for all three explanatory variables indicate that they are statistically significant. 

As the humidity and windspeed increase, the difference in temperate decreases (apparent - actual). 

Humidity had a much stronger negative correlation with temperature than with temperature difference. 

In [0]:
#create new variable
df['humidity_windspeed'] = df.humidity * df.windspeed
y = df['temp_diff']
X = df[['humidity', 'windspeed', 'humidity_windspeed']]

In [24]:
X = sm.add_constant(X)

results = sm.OLS(y, X).fit()

results.summary()

/usr/local/lib/python3.6/dist-packages/numpy/core/fromnumeric.py:2495: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              temp_diff   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Tue, 05 Nov 2019   Prob (F-statistic):               0.00
Time:                        14:19:23   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                  0.0839      0.033      2.511      0.012       0.018       0.149
humidity               0.1775      0.043      4.133      0.000       0.093       0.262
windspeed              0.0905      0.002     36.797      0.000       0.086       0.095
humidity_windspeed    -0.2971      0.003    -88.470      0.000      -0.304      -0.291
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.265
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                          -0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The coefficients are statistically significant since pvalue is 0. The coefficients tell us that as humudity and windspeed increase, humidty_windspeed do as well. 

We can see that the humidity and windspeed interaction has a negative coefficient but humidity and windspeed by themselves have a positive coefficient. 

This is because the windspeed mitigates the effects of humidity on the target. 

# 3. House prices model

## Recreate dataframe

In [0]:
#Make SQL Queary
postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'houseprices'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

df = pd.read_sql_query('select * from houseprices',con=engine)

# no need for an open connection, as we're only doing a single query
engine.dispose()

In [0]:
#Drop columns that have a large number of null values (over 30%)
df1 = df.drop(columns = ['alley', 'fireplacequ', 'poolqc', 'fence', 'miscfeature'])

In [0]:
#make a copy of df1
df2 = df1.copy()

#interpolate based on the year
year_list = list(df1['yearbuilt'].unique())
for year in year_list:
    df2.loc[df1['yearbuilt']==year] = df2.loc[df1['yearbuilt']==year].interpolate()

# drop missing values that are remaining
df2.dropna(inplace = True)

In [0]:
#update data types
non_numeric = df1.select_dtypes(include = 'object')
numeric = df1.select_dtypes(include = ['float64', 'int64'])

In [0]:
#pick top 5 numeric:
df3 = df2.loc[:, ['overallqual', 'grlivarea', 'garagecars', 'garagearea', 'totalbsmtsf', 'saleprice']]

In [0]:
#Choose Central Air and External Quality as the features I'll use for the model
df4 = df2.loc[:, ['exterqual', 'centralair']]

#One Hot encoding of categorical variables
df5 = pd.get_dummies(df4, columns = ['exterqual', 'centralair'])

In [0]:
#Combine Dataframes
df3
df5
model_df = pd.concat([df3, df5], axis = 1)

## Create model

In [0]:
#Separate target variable from rest of data
X = model_df.drop(columns = 'saleprice')
y = model_df['saleprice']

#split train_test
X_train, X_test, y_train, y_test = train_test_split(X, y)

In [33]:
X = sm.add_constant(X)

results = sm.OLS(y, X).fit()

results.summary()

/usr/local/lib/python3.6/dist-packages/numpy/core/fromnumeric.py:2495: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.775
Model:                            OLS   Adj. R-squared:                  0.773
Method:                 Least Squares   F-statistic:                     500.0
Date:                Tue, 05 Nov 2019   Prob (F-statistic):               0.00
Time:                        14:46:30   Log-Likelihood:                -15757.
No. Observations:                1318   AIC:                         3.153e+04
Df Residuals:                    1308   BIC:                         3.159e+04
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const        -4.127e+04   4923.196     -8.383      0.000   -5.09e+04   -3.16e+04
overallqual   1.884e+04   1335.192     14.107      0.000    1.62e+04    2.15e+04
grlivarea       47.6567      2.612     18.249      0.000      42.533      52.780
garagecars    1.645e+04   3147.805      5.227      0.000    1.03e+04    2.26e+04
garagearea      13.4964     10.494      1.286      0.199      -7.091      34.084
totalbsmtsf     26.2237      3.274      8.011      0.000      19.802      32.646
exterqual_Ex  3.735e+04   6542.041      5.709      0.000    2.45e+04    5.02e+04
exterqual_Fa -3.957e+04   1.15e+04     -3.435      0.001   -6.22e+04    -1.7e+04
exterqual_Gd -1.261e+04   4084.879     -3.086      0.002   -2.06e+04   -4591.762
exterqual_TA -2.645e+04   3568.598     -7.411      0.000   -3.34e+04   -1.94e+04
centralair_N -3.273e+04   3714.024     -8.812      0.000      -4e+04   -2.54e+04
centralair_Y -8544.2175   3417.795     -2.500      0.013   -1.52e+04   -1839.257
==============================================================================
Omnibus:                      603.343   Durbin-Watson:                   1.972
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            68036.538
Skew:                          -1.125   Prob(JB):                         0.00
Kurtosis:                      38.126   Cond. No.                     5.96e+19
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 1.55e-30. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

All of the features are statitically signficant (p<.05) except for garagearea (which has a pvalue of .199).

## Remove insignificant features

In [0]:
#Separate target variable from rest of data
X = model_df.drop(columns = ['saleprice', 'garagearea'])
y = model_df['saleprice']

In [37]:
X = sm.add_constant(X)

results = sm.OLS(y, X).fit()

results.summary()

/usr/local/lib/python3.6/dist-packages/numpy/core/fromnumeric.py:2495: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.775
Model:                            OLS   Adj. R-squared:                  0.773
Method:                 Least Squares   F-statistic:                     562.0
Date:                Tue, 05 Nov 2019   Prob (F-statistic):               0.00
Time:                        14:52:41   Log-Likelihood:                -15758.
No. Observations:                1318   AIC:                         3.153e+04
Df Residuals:                    1309   BIC:                         3.158e+04
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const        -4.095e+04   4918.147     -8.327      0.000   -5.06e+04   -3.13e+04
overallqual   1.876e+04   1334.287     14.061      0.000    1.61e+04    2.14e+04
grlivarea       47.9542      2.602     18.430      0.000      42.850      53.059
garagecars    1.941e+04   2147.534      9.040      0.000    1.52e+04    2.36e+04
totalbsmtsf     27.1598      3.192      8.507      0.000      20.897      33.423
exterqual_Ex  3.745e+04   6543.184      5.724      0.000    2.46e+04    5.03e+04
exterqual_Fa -3.856e+04   1.15e+04     -3.354      0.001   -6.11e+04    -1.6e+04
exterqual_Gd   -1.3e+04   4074.614     -3.189      0.001    -2.1e+04   -5002.082
exterqual_TA -2.685e+04   3556.073     -7.549      0.000   -3.38e+04   -1.99e+04
centralair_N -3.267e+04   3714.701     -8.795      0.000      -4e+04   -2.54e+04
centralair_Y -8280.1366   3412.473     -2.426      0.015    -1.5e+04   -1585.622
==============================================================================
Omnibus:                      585.883   Durbin-Watson:                   1.972
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            63696.739
Skew:                          -1.071   Prob(JB):                         0.00
Kurtosis:                      36.990   Cond. No.                     6.69e+19
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 1.15e-30. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

Dropping garagearea had a small impact on a few features. It made garagecars have a higher coefficient. The least prominent features are totalbsmtsf and grlivarea. The other features like overall quality, garage cars, external quality, and central air all have a much larger impact on saleprice. 